pintracker/statelesstracker/statelesstracker.go

```diff
@@ -0,0 +1,380 @@
+				spt.optracker.UpdateOperationPhase(p.Cid, optracker.PhaseInProgress)
+				spt.unpin(p)
+			}
+		case <-spt.ctx.Done():
+			return
+		}
+	}
+}
+
+// SetClient makes the StatelessPinTracker ready to perform RPC requests to
+// other components.
+func (spt *StatelessPinTracker) SetClient(c *rpc.Client) {
+	spt.rpcClient = c
+	spt.rpcReady <- struct{}{}
+}
+
+// Shutdown finishes the services provided by the StatelessPinTracker
+// and cancels any active context.
+func (spt *StatelessPinTracker) Shutdown() error {
+	spt.shutdownMu.Lock()
+	defer spt.shutdownMu.Unlock()
+
+	if spt.shutdown {
+		logger.Debug("already shutdown")
+		return nil
+	}
+
+	logger.Info("stopping StatelessPinTracker")
+	spt.cancel()
+	close(spt.rpcReady)
+	spt.wg.Wait()
+	spt.shutdown = true
+	return nil
+}
+
+// Track tells the StatelessPinTracker to start managing a Cid,
+// possibly triggering Pin operations on the IPFS daemon.
+func (spt *StatelessPinTracker) Track(c api.Pin) error {
+	logger.Debugf("tracking %s", c.Cid)
+	if ptutil.IsRemotePin(c, spt.peerID) {
+		if spt.Status(c.Cid).Status == api.TrackerStatusPinned {
+			spt.optracker.TrackNewOperation(
+				spt.ctx,
+				c.Cid,
+				optracker.OperationUnpin,
+			)
+			spt.unpin(c)
+		}
+		return nil
+	}
+
+	if opc, ok := spt.optracker.Get(c.Cid); ok {
+		if opc.Op == optracker.OperationUnpin {
+			switch opc.Phase {
+			case optracker.PhaseQueued:
+				spt.optracker.Finish(c.Cid)
+				return nil
+			case optracker.PhaseInProgress:
+				spt.optracker.Finish(c.Cid)
+				// NOTE: this may leave the api.PinInfo in an error state
+				// so a pin operation needs to be run on it (same as Recover)
+			}
+		}
+	}
+
+	spt.optracker.TrackNewOperation(spt.ctx, c.Cid, optracker.OperationPin)
+
+	select {
+	case spt.pinCh <- c:
+	default:
+		err := errors.New("pin queue is full")
+		spt.optracker.Finish(c.Cid)
+		logger.Error(err.Error())
+		return err
+	}
+	return nil
+}
+
+// Untrack tells the StatelessPinTracker to stop managing a Cid.
+// If the Cid is pinned locally, it will be unpinned.
+func (spt *StatelessPinTracker) Untrack(c *cid.Cid) error {
+	logger.Debugf("untracking %s", c)
+	if opc, ok := spt.optracker.Get(c); ok {
+		if opc.Op == optracker.OperationPin {
+			spt.optracker.Finish(c) // cancel it
+
+			switch opc.Phase {
+			case optracker.PhaseQueued:
+				return nil
+			case optracker.PhaseInProgress:
+				// continues below to run a full unpin
+			}
+		}
+	}
+
+	spt.optracker.TrackNewOperation(spt.ctx, c, optracker.OperationUnpin)
+
+	select {
+	case spt.unpinCh <- api.PinCid(c):
+	default:
+		err := errors.New("unpin queue is full")
+		spt.optracker.Finish(c)
+		logger.Error(err.Error())
+		return err
+	}
+	return nil
+}
+
+// StatusAll returns information for all Cids pinned to the local IPFS node.
+func (spt *StatelessPinTracker) StatusAll() []api.PinInfo {
+	// get statuses from ipfs node first
+	localpis, _ := spt.ipfsStatusAll()
```

Comment:
ipfsStatusAll receives a map, which is converted to a slice, which is here converted again to a map, which is then converted to a slice. There should be just one map->slice operation.